### PR TITLE
add mobilenet and v2 backbones

### PR DESCRIPTION
**Pull request by Maxence1**

### CODE REVIEW

File: models/backbone/__init__.py [hunk 1]
```diff
@@ -6,4 +6,4 @@
 from .se_resnext import SEResNeXt
 
 __all__ = ['ResNet', 'ResNeXt', 'ShuffleNet', 'ShuffleNetV2',
-           'SEResNet', 'SEResNeXt']
+           'SEResNet', 'SEResNeXt', 'MobileNets' 'MobileNetV2']
```
Comment (wsnedy) [inline models/backbone/__init__.py]: Just use 'MobileNet', remove 's'

File: models/backbone/mobilenet.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
+import logging
+import torch.nn as nn
+
+
+from ..utils import conv1x1_group, conv3x3_group, \
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: too much blank line, remove one

File: models/backbone/mobilenet.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+class Conv_dw_pw(nn.Module):
+
+    def __init__(self,
+                 inplanes,
+                 planes,
+                 stride=1,
+                 dilation=1,
+                 use_gn=False,
+                 widthmultiplier=1):
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: 'widthmultiplier' is too long, use a short name 'width_mult' instead

File: models/backbone/mobilenet.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        x = self.dw_conv(x)
+        norm = getattr(self, self.norm_names[0])
+        x = norm(x)
+        x = nn.ReLU(x)
+
+        # pw
+        x = self.pw_conv(x)
+        norm = getattr(self, self.norm_names[0])
+        x = norm(x)
+        x = nn.ReLU(x)
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: 1. add the attribute 'stride', 'dilation', 'use_gn' as the block in ResNet
2. use 'norm1' and 'norm2' instead for two norm layer in 'forward'.
3. add a 'relu' layer in '__init__' function, you can reuse it in 'forward'. 
`
def __init__(self, ...):
      self.relu = nn.ReLU(inplace=True)

def forward(self, ...):
      x = norm1(x)
      x = self.relu(x)
      ....
     x = norm2(x)
     x = self.relu(x)
     ...
`

File: models/backbone/mobilenet.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                 widthmultiplier=1):
+    layers = [
+        block(
+            inplanes,
+            planes,
+            stride,
+            dilation,
+            use_gn)
+    ]
+    return nn.Sequential(*layers)
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: this function need given `blocks` to indicate the number of blocks in this stage, and `layers` should be a list that contain `blocks` of block

File: models/backbone/mobilenet.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+
+class MobileNets(nn.Module):
+    r"""PyTorch implementation of the MobileNets architecture
+    <https://arxiv.org/abs/1704.04861>`_.
+    Model has been designed to work on either ImageNet or CIFAR-10
+    Args:
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: do not need `r`, clean the `https` by removing '<', '>', etc., add a blank line here to seperate the `Args`

File: models/backbone/mobilenet.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        num_classes (int): 1000 for ImageNet, 10 for CIFAR-10
+        large_img (bool): True for ImageNet, False for CIFAR-10
+    """
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: the `Args` here should be the description of paramters in `__init__` function, please update it

File: models/backbone/mobilenet.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    arch_settings = [[(32, 64, 1, 1)],
+                     [(64, 128, 2, 1), (128, 128, 1, 1)],
+                     [(128, 256, 2, 1), (256, 256, 1, 1)],
+                     [(256, 512, 2, 1), (512, 512, 1, 1),
+                      (512, 512, 1, 1), (512, 512, 1, 1),
+                      (512, 512, 1, 1), (512, 512, 1, 1)],
+                     [(512, 1024, 2, 1), (1024, 1024, 1, 1)]]
+
+    def __init__(self,
+                 num_stages=13,
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: There are only 5 stages in MobileNet, and your `arch_settings` also shows 5 stages

File: models/backbone/mobilenet.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    """
+
+    # (inplanes, planes, stride, dilation)
+    arch_settings = [[(32, 64, 1, 1)],
+                     [(64, 128, 2, 1), (128, 128, 1, 1)],
+                     [(128, 256, 2, 1), (256, 256, 1, 1)],
+                     [(256, 512, 2, 1), (512, 512, 1, 1),
+                      (512, 512, 1, 1), (512, 512, 1, 1),
+                      (512, 512, 1, 1), (512, 512, 1, 1)],
+                     [(512, 1024, 2, 1), (1024, 1024, 1, 1)]]
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: I recommend to change the `arch_setting` as follow:
```
arch_settings = {
0.5: ((32, 64, 128, 256, 512), (1, 2, 2, 6, 2))
1.0: ((64, 128, 256, 512, 1024), (1, 2, 2, 6, 2))
.....
}
```

File: models/backbone/mobilenet.py [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+
+    def __init__(self,
+                 num_stages=13,
+                 out_indices=(0, 1, 2, 3, 4),
+                 forzen_stages=-1,
+                 use_gn=False,
+                 bn_eval=True,
+                 bn_forzen=False,
+                 widthmultiplier=1):
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: add `strides` and `dilations` in the `__init__` function, refer to ShuffleNetv2

File: models/backbone/mobilenet.py [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            for j, (inplanes, planes, stride, dilation) in enumerate(
+                    stage_arch):
+                mobilev1_layer = _make_layers(
+                    Conv_dw_pw,
+                    inplanes,
+                    planes,
+                    stride,
+                    dilation,
+                    use_gn,
+                    widthmultiplier)
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: move the loop into `_make_layers` function, and rewrite this snippet

File: models/backbone/mobilenet.py [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+                elif isinstance(m, (nn.BatchNorm2d, nn.GroupNorm)):
+                    constant_init(m, 1)
+        else:
+            raise TypeError('pretrained must be a str or None')
+
+    def forward(self, x):
+        x = self.conv1(x)
+        norm1 = getattr(self, self.norm1_name)
+        x = norm1(x)
+        x = nn.ReLU(x)
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: also add a `relu` layer in `__init__`

File: models/backbone/mobilenetv2.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+    def forward(self, x):
+        out = self.conv1(x)
+        norm1 = getattr(self, self.norm_names[0])
+        out = norm1(out)
+        out = nn.ReLU6(out)
+
+        out = self.conv2(out)
+        norm2 = getattr(self, self.norm_names[1])
+        out = norm2(out)
+        out = nn.ReLU6(out)
```
Comment (wsnedy) [inline models/backbone/mobilenetv2.py]: add a `relu6` layer in `__init__`

File: models/backbone/mobilenetv2.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+
+class LinearBottleNeck(nn.Module):
+    def __init__(self,
+                 inplanes,
+                 outplanes,
+                 expansion=6,
+                 stride=1,
+                 dilation=1,
+                 use_gn=False,
+                 downsample=None):
```
Comment (wsnedy) [inline models/backbone/mobilenetv2.py]: use `shortcut` instead of `downsample`

File: models/backbone/mobilenetv2.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+                 expansion=6,
+                 stride=1,
+                 dilation=1,
+                 use_gn=False):
+    shortcut = None
+    if stride != 1 or inplanes != outplanes:
+        shortcut = nn.Sequential(
+            conv1x1_group(inplanes, outplanes),
+            norm_layer(outplanes, use_gn)
+        )
```
Comment (wsnedy) [inline models/backbone/mobilenetv2.py]: This is not correct, remove the `shortcut`

File: models/backbone/mobilenetv2.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+        out = self.conv3(out)
+        norm3 = getattr(self, self.norm_names[2])
+        out = norm3(out)
+
+        if self.stride == 1 & self.inplanes == self.planes:
+            out += self.shortcut(x)
+        return out
+
+
+def _make_layers(block,
```
Comment (wsnedy) [inline models/backbone/mobilenetv2.py]: this function is not correct, need rewrite according to the `figure4` in paper

File: models/backbone/mobilenetv2.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+                elif isinstance(m, (nn.BatchNorm2d, nn.GroupNorm)):
+                    constant_init(m, 1)
+        else:
+            raise TypeError('pretrained must be a str or None')
+
+    def forward(self, x):
+        x = self.conv1(x)
+        norm1 = getattr(self, self.norm1_name)
+        x = norm1(x)
+        x = nn.ReLU6(x)
```
Comment (wsnedy) [inline models/backbone/mobilenetv2.py]: add `relu` layer

File: models/backbone/mobilenetv2.py [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+                  stride=1,
+                  dilation=dilation,
+                  use_gn=use_gn,
+                  shortcut=shortcut)
+        )
+    return nn.Sequential(*layers)
+
+
+class MobileNetV2(nn.Module):
+    # (expansion, out_planes, num_blocks, stride, dilation)
```
Comment (wsnedy) [inline models/backbone/mobilenetv2.py]: add docstring for MobileNetV2

File: models/backbone/mobilenet.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+        self.use_gn = use_gn
+
+        self.inplanes = round(32 * width_multi)
+        self.conv1 = conv3x3_group(3, round(32 * width_multi), stride=2)
+        self.norm1_name = 'bn1' if not use_gn else 'gn1'
+        self.add_module(self.norm1_name,
+                        norm_layer(round(32 * width_multi), use_gn))
+        self.relu = nn.ReLU(inplace=True)
+
+        stage_blocks, num_blocks = self.arch_settings[width_multi]
```
Comment (wsnedy) [inline models/backbone/mobilenet.py]: I think here should be:
```
stage_outplanes, stage_blocks = self.arch_settings[width_multi]
```
And the use of `stage_blocks` in this implementation has minor bugs, you should fix it
